File: golang/three_sum.go
func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	result := [][]int{}

	// find a start point, then run two sum with remaining
	for i := 0; i < len(nums)-2; i++ {
		// impossible to have sum of 0 bc sorted
		if nums[i] > 0 {
			break
		}

		// avoid duplicates
		if i == 0 || nums[i] != nums[i-1] {
			start, end, sum := i+1, len(nums)-1, -nums[i] // need nums[start]+nums[end] = -nums[i]

			for start < end {
				if nums[start]+nums[end] == sum {
					res := []int{nums[i], nums[start], nums[end]}
					result = append(result, res)

					// skip duplicates
					for start < end && nums[start] == nums[start+1] {
						start++
					}
					for start < end && nums[end] == nums[end-1] {
						end--
					}
					start++
					end--
				} else if nums[start]+nums[end] < sum {
					start++
				} else if nums[start]+nums[end] > sum {
					end--
				}
			}
		}
	}

	return result
}

File: typescript/three_sum.ts
function threeSum(nums: number[]): number[][] {
	nums.sort((x, y) => x - y)

	const result = []

	for (let i = 0; i < nums.length - 2; i++) {
		// since sorted, it is impossible to sum 0 if pivot >= 0
		if (nums[i] > 0) {
			continue
		}

		// avoid duplicate pivot
		if (i == 0 || (i != 0 && nums[i] != nums[i - 1])) {
			let start = i + 1
			let end = nums.length - 1
			let target = -nums[i]

			// perform two sum with start and end
			while (start < end) {
				let sum = nums[start] + nums[end]
				if (sum == target) {
					result.push([ nums[i], nums[start], nums[end] ])

					start++
					while (nums[start] == nums[start - 1]) {
						start++
					}
					end--
					while (nums[end] == nums[end + 1]) {
						end--
					}
				} else if (sum < target) {
					start++
					while (nums[start] == nums[start - 1]) {
						start++
					}
				} else if (sum > target) {
					end--
					while (nums[end] == nums[end + 1]) {
						end--
					}
				}
			}
		}
	}

	return result
}
